Framework/DataHandling/src/LoadSampleShape.cpp

```diff
@@ -0,0 +1,222 @@
+
+  if (readSTLLine(file, "facet") && readSTLLine(file, "outer loop")) {
+    bool ok = (readSTLVertex(file, v1) && readSTLVertex(file, v2) &&
+               readSTLVertex(file, v3));
+    if (!ok) {
+      throw std::runtime_error("Error on reading STL triangle");
+    }
+  } else {
+    return false; // End of file
+  }
+  return readSTLLine(file, "endloop") && readSTLLine(file, "endfacet");
+}
+
+// Adds vertex to list if distinct and returns index to vertex added or equal
+uint16_t addSTLVertex(V3D &vertex, std::vector<V3D> &vertices) {
+  for (uint16_t i = 0; i < vertices.size(); ++i) {
+    if (areEqualVertices(vertex, vertices[i])) {
+      return i;
+    }
+  }
+  vertices.push_back(vertex);
+  uint16_t index = static_cast<uint16_t>(vertices.size() - 1);
+  if (index != vertices.size() - 1) {
+    throw std::runtime_error("Too many vertices in solid");
+  }
+  return index;
+}
+
+std::unique_ptr<MeshObject> readSTLMeshObject(std::ifstream &file) {
+  std::vector<uint16_t> triangleIndices;
+  std::vector<V3D> vertices;
+  V3D t1, t2, t3;
+
+  while (readSTLTriangle(file, t1, t2, t3)) {
+    // Add triangle if all 3 vertices are distinct
+    if (!areEqualVertices(t1, t2) && !areEqualVertices(t1, t3) &&
+        !areEqualVertices(t2, t3)) {
+      triangleIndices.push_back(addSTLVertex(t1, vertices));
+      triangleIndices.push_back(addSTLVertex(t2, vertices));
+      triangleIndices.push_back(addSTLVertex(t3, vertices));
+    }
+  }
+  // Use efficient constructor of MeshObject
+  std::unique_ptr<MeshObject> retVal = std::unique_ptr<MeshObject>(
+      new MeshObject(std::move(triangleIndices), std::move(vertices),
+                     Mantid::Kernel::Material()));
+  return retVal;
+}
+
+std::unique_ptr<Geometry::MeshObject> readSTLSolid(std::ifstream &file,
+                                                   std::string &name) {
+  // Read Solid name
+  // We expect line after trimming to be "solid "+name.
+  std::string line;
+  if (getline(file, line)) {
+    boost::trim(line);
+    if (line.size() < 5 || line.substr(0, 5) != "solid") {
+      throw std::runtime_error("Expected start of solid");
+    } else {
+      name = line.substr(6, std::string::npos);
+    }
+    // Read Solid shape
+    return readSTLMeshObject(file);
+  }
+  return nullptr;
+}
+
+} // end anonymous namespace
+
+void LoadSampleShape::init() {
+  auto wsValidator = boost::make_shared<API::InstrumentValidator>();
+  ;
+
+  // input workspace
+  declareProperty(
+      make_unique<WorkspaceProperty<>>("InputWorkspace", "", Direction::Input,
+                                       wsValidator),
+      "The name of the workspace containing the instrument to add the shape");
+
+  // shape file
+  const std::vector<std::string> extensions{".stl"};
+  declareProperty(
+      make_unique<FileProperty>("Filename", "", FileProperty::Load, extensions),
+      "The path name of the file containing the shape.");
```

Comment:
Full stop at the end of this property description, but not the other two?  Consistency